postmortems/2024-03-22_azure_dev_outage.md

```diff
@@ -0,0 +1,110 @@
+* 2024/03/26 Tuesday 17:46 PO emails two individuals per advice of IT Service Desk 
+* 2024/03/26 Tuesday 21:44 Line manager indicates a change request which goes through the Technical Review Board and Change Approval Board will be required to implement the change once the secret is received 
+* 2024/03/26 Tuesday 22:51 Ticket re access is first round approved 
+* 2024/03/27 Wednesday 08:20 MP receives the new secret details
+* 2024/03/27 Wednesday 09:47 Ticket re access is second round approved 
+* 2024/03/27 Wednesday 09:51 Remainder of team becomes aware of new secret details
+* 2024/03/27 Wednesday 10:50 Other internal user speaks to line manager for approval to upload the new client secret – indicates we need authorisation not to go through a change request 
+* 2024/03/27 Wednesday 11:53 MP confirms he has access to his client account again and is able to login 
+* 2024/03/27 Wednesday 13:41 PO was successfully added as an owner to the app registration
+* 2024/03/27 Wednesday 14:26 MP able to add PO as an owner to the DEV environment app registration 
+* 2024/03/28 Thursday 13:30 Internal user and MP attended CAB and got approval to upload secret 
+* 2024/03/28 Thursday 15:34 app is back online 
+
+
+## Why
+
+### Why were the Azure Monitor Alerts which should have indicated the problem not more quickly responded to? 
+
+While we received a number of Azure Monitor Alerts to the Support inbox  – alerts are habitually received for a known issue related to the web service sending out a 500 error when a connection is dropped instead of handling it. We use async handling rather than having a public blob storage, which makes the server report an error when connection is dropped that it shouldn’t. The Azure Monitor Alerts received from Friday 22 March 2024 through to Monday 25 March 2024 before first internal user reported the issue were not properly examined as they were inappropriately assumed to have been caused by the known issue, rather than a new problem.  
+
+In addition to this, the first alert was received at 14:38 on the Friday, but only forwarded to MP & AL at 10:38 on Monday.  
+
+### Why was the secret allowed to expire? 
+
+1. No system reminder to renew secret 
+The client Azure system does not automatically send any notifications for impending expiry of secret – based on content in Azure forums it seems to be a common problem for Azure secrets beyond this client environment. There isn’t an obvious place for us to set these up for ourselves, but we have also never asked about it. We intend to look into some scripts that we could potentially run that would enable automated reminders for the secret expiry.  
+
+Given that one third party supplier sends an excel spreadsheet monthly of Service Principals expiring, we believe this must be a widespread problem. 
```

Comment:
I'm not sure this being a widespread problem is that relevant? Yes, Azure secret management is bad, but it's what we have to work with. It's worth having a note that this isn't a specific feature of the client Azure environment and is instead just a shortcoming of Azure in general, but in so far as this specific outage was concerned all that matters is that:

- Azure has no functionality to automatically alert app registration/service principal owners that an attached secret is about to expire.
- Therefore we need to make our own alerts or reminders, automated or otherwise.
- For this specific secret, we did not do this.